contracts/examples/lottery-esdt/src/lottery.rs

```diff
@@ -277,33 +274,161 @@ pub trait Lottery {
         } else {
             info.prize_distribution.len()
         };
-        let total_prize = info.prize_pool.clone();
-        let winning_tickets = self.get_distinct_random(1, total_tickets, total_winning_tickets);
-
-        // distribute to the first place last. Laws of probability say that order doesn't matter.
-        // this is done to mitigate the effects of BigUint division leading to "spare" prize money being left out at times
-        // 1st place will get the spare money instead.
-        for i in (1..total_winning_tickets).rev() {
-            let winning_ticket_id = winning_tickets[i];
-            let winner_address = ticket_holders_mapper.get(winning_ticket_id);
-            let prize = self.calculate_percentage_of(
-                &total_prize,
-                &BigUint::from(info.prize_distribution.get(i)),
-            );
 
+        self.total_winning_tickets(lottery_name)
+            .set(total_winning_tickets);
+        self.index_last_winner(lottery_name).set(1);
+    }
+
+    fn distribute_prizes(&self, lottery_name: &ManagedBuffer) -> AwardingStatus {
+        let mut info = self.lottery_info(lottery_name).get();
+        let ticket_holders_mapper = self.ticket_holders(lottery_name);
+        let total_tickets = ticket_holders_mapper.len();
+
+        let mut index_last_winner = self.index_last_winner(lottery_name).get();
+        let total_winning_tickets = self.total_winning_tickets(lottery_name).get();
+        require!(
+            index_last_winner <= total_winning_tickets,
+            "Awarding has ended"
+        );
+
+        let mut iterations = 0;
+        while index_last_winner <= total_winning_tickets && iterations < MAX_OPERATIONS {
+            let rand_index = self.get_distinct_random(index_last_winner, total_tickets, 1)[0];
```

Comment:
This way of generating random numbers is terribly inefficient if all you want is _one_ number. If this is the only place where you use that function, simply rewrite it to generate one number in range `(index_last_winner, total_tickets)`